File: Python3/Longest Common Prefix.py
# 36ms, 100.0%
class Solution:
    def longestCommonPrefix(self, strs):
        if not strs: return ""
        if len(strs) == 1: return strs[0]
        beg, mid, end = 0, 0, len(strs[0]) - 1
        while beg <= end:
            mid = (beg + end + 1) // 2
            for s in strs:
                if strs[0][:mid + 1] != s[:mid + 1]:
                    end = mid - 1
                    break
            else:
                if end == beg: return strs[0][:mid + 1]
                else:          beg = mid
        return ""
        

File: Java/Longest Common Prefix.java
// 0ms, 100.00%; 37.3MB, 99.42%
class Solution {
    public static String longestCommonPrefix(final String[] strs) {
        if (strs.length == 0) {
            return "";
        }
        int index = strs[0].length();
        for (String str : strs) {
            if (str.length() < index) {
                index = str.length();
            }
        }
        for (int i = 0, j; i < index; ++i) {
            for (j = 1; j < strs.length; ++j) {
                if (strs[0].charAt(i) != strs[j].charAt(i)) {
                    break;
                }
            }
            if (j < strs.length) {
                return strs[0].substring(0, i);
            }
        }
        return strs[0].substring(0, index);
    }
}
